tests/test_cli.py

```diff
@@ -0,0 +1,199 @@
+    """
+
+    # Temporary surrogate `package.json` file.
+    package_json = {"scripts": {"build": "echo Hotzenplotz"}}
+    package_json_file = tmp_path / "package.json"
+    package_json_file.write_text(json.dumps(package_json))
+
+    # Invoke `responder build`.
+    stdout, stderr = responder_build(tmp_path, capfd)
+    assert "Hotzenplotz" in stdout
+
+
+def test_cli_build_missing_package_json(capfd, tmp_path):
+    """
+    Verify `responder build`, while `package.json` file is missing.
+    """
+
+    # Invoke `responder build`.
+    stdout, stderr = responder_build(tmp_path, capfd)
+    assert "npm error code ENOENT" in stderr
+    assert "Could not read package.json" in stderr
+    assert "Error: ENOENT: no such file or directory" in stderr
+
+
+@pytest.mark.parametrize(
+    "invalid_content,npm_error,expected_error",
+    [
+        ("foobar", "code EJSONPARSE", ["is not valid JSON", "Failed to parse JSON data"]),
+        ("{", "code EJSONPARSE", "Unexpected end of JSON input"),
+        ('{"scripts": }', "code EJSONPARSE", "Unexpected token"),
+        (
+            '{"scripts": null}',
+            "Cannot convert undefined or null to object",
+            "scripts.build script not found",
+        ),
+        ('{"scripts": {"build": null}}', "Missing script", '"build"'),
+        ('{"scripts": {"build": 123}}', "Missing script", '"build"'),
+    ],
+    ids=[
+        "invalid_json_content",
+        "incomplete_json",
+        "syntax_error",
+        "null_scripts",
+        "missing_script_null",
+        "missing_script_number",
+    ],
+)
+def test_cli_build_invalid_package_json(
+    capfd, tmp_path, invalid_content, npm_error, expected_error
+):
+    """
+    Verify `responder build` using an invalid `package.json` file.
+    """
+
+    # Temporary surrogate `package.json` file.
+    package_json_file = tmp_path / "package.json"
+    package_json_file.write_text(invalid_content)
+
+    # Invoke `responder build`.
+    stdout, stderr = responder_build(tmp_path, capfd)
+    assert f"npm error {npm_error}" in stderr
+    assert any(item in stderr for item in to_list(expected_error))
+
+
+def test_cli_run(capfd):
+    """
+    Verify that `responder run` works as expected.
+    """
+
+    # Sanity checks.
+    target = Path("examples") / "helloworld.py"
+    if not target.is_file():
+        raise FileNotFoundError(f"Example file not found: {target}")
+
+    # Invoke `responder run`.
+    # Start a Responder service instance in the background, using its CLI.
+    # Make it terminate itself after serving one HTTP request.
+    server = ResponderServer(target=str(target), port=random_port(), limit_max_requests=1)
+    try:
+        server.start()
+        # Can't use `wait_server` here, because probing
+        # requests would terminate the server.
+        # wait_server(server.port)
+        # Need to fall back to naive sleeping for now.
+        time.sleep(1)
+        response = requests.get(
```

Comment:
Not sure that the condition `if "Application startup complete" in capfd.readouterr().err:` is applicable?

Reply:
https://github.com/kennethreitz/responder/blob/3e2a088c0b7202d326f62281ac4068f349a6b7ba/tests/test_cli.py#L176-L180

> Since we can't send HTTP requests for probing due to the single-request limit [...]

Thank you for taking the inline comments into consideration.

> [...] an alternative is to check if the server's port is open.

Thanks a stack for providing an alternative implementation. We will try if that works better and more reliable on a subsequent iteration.

Reply:
This suggestion, amongst other changes, seem to to work well now to make the test suite on CI/GHA more robust. Thanks again.